### PR TITLE
Closes CPPL-281

### PR DESCRIPTION
**Pull request by sorbeee**
Implemented PeriodicIncomeRepository

### CODE REVIEW

File: src/Server/Tests/MockDbFacade.h [hunk 1]
```diff
@@ -65,4 +65,27 @@ class MockDbFacade : public db::IDbFacade
...
+	MOCK_METHOD(std::vector<db::PeriodicIncome>, GetAllPeriodicIncomes, (db::IdType user_id), (override));
+	MOCK_METHOD(bool, UpdatePeriodicIncome, (const db::PeriodicIncome& income), (override));
+	MOCK_METHOD(bool, RemovePeriodicIncome, (db::IdType id), (override));
+	MOCK_METHOD(bool, CanUserEditPeriodicIncome, (db::IdType user_id, db::IdType income_id), (override));
+	MOCK_METHOD(std::vector<db::PeriodType>, GetAllPeriodTypes, (), (override));
+	MOCK_METHOD(bool, CanGeneratePeriodicIncome, (db::IdType user_id, db::IdType periodic_id), (override));
+	MOCK_METHOD(bool, UpdateAddNextPeriodicIncome, (const db::PeriodicIncome& income), (override));
+	/*
+	 *
+	virtual std::optional<IdType> AddPeriodicIncome(const PeriodicIncome& income) = 0;
```
Comment (iodream) [inline src/Server/Tests/MockDbFacade.h]: are these here necessary?

**Review by iodream — APPROVED** [review]

File: src/Server/libdal/Repositories/PeriodicIncomeRepository/PeriodicIncomeRepository.cpp [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+			" FROM  " + periodicIncome::TABLE_NAME +
+			" WHERE " + periodicIncome::ID + " = " + w.quote(income.id));
+
+		if (result.empty())
+		{
+			return false;
+		}
+
+		w.exec0(
+				"UPDATE " + periodicIncome::TABLE_NAME +
```
Comment (sorbeee) [inline src/Server/libdal/Repositories/PeriodicIncomeRepository/PeriodicIncomeRepository.cpp]: Fix this

File: src/Server/libdal/Repositories/PeriodicIncomeRepository/PeriodicIncomeRepository.cpp [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+			" FROM  " + periodicIncome::TABLE_NAME +
+			" WHERE " + periodicIncome::ID + " = " + w.quote(income.id));
+
+		if (result.empty())
+		{
+			return false;
+		}
+
+		w.exec0(
+			"UPDATE " + periodicIncome::TABLE_NAME +
```
Comment (sorbeee) [inline src/Server/libdal/Repositories/PeriodicIncomeRepository/PeriodicIncomeRepository.cpp:289]: There too

File: SQLScripts/initDB.sql [hunk 4]
```diff
@@ -15,6 +15,26 @@ create table IncomeCategory
...
+
+create table PeriodicIncome
+(
+    id bigint generated by default as identity primary key,
+	name varchar(30) not null,
+	userId bigint not null references User_(id) on delete cascade,
+	amount bigint not null,
+	categoryId bigint not null references IncomeCategory(id),
+
+    periodId bigint not null references PeriodType(id),
```
Comment (yaroslav-kernitskyi) [inline SQLScripts/initDB.sql]: Spaces instead of tab

File: SQLScripts/initDB.sql [hunk 5]
```diff
@@ -15,6 +15,26 @@ create table IncomeCategory
...
+
+create table PeriodType
+(
+    id bigint generated by default as identity primary key,
+	name varchar(20) not null
+);
+
+create table PeriodicIncome
+(
+    id bigint generated by default as identity primary key,
```
Comment (yaroslav-kernitskyi) [inline SQLScripts/initDB.sql]: Spaces

File: src/Server/Handlers/Product/GetRecommendationProductHandler.cpp [hunk 6]
```diff
@@ -27,22 +27,22 @@ Net::Response GetRecommendationProductHandler::AuthHandle(const Net::Request& re
 			"\" can't get recommendation product for user with id \"" + std::to_string(user_id) + "\"");
 	}
 
-	db::Product db_product;
+	std::optional<db::Product> db_product;
 	try
 	{
-		db_product = m_facade->GetRecommendation(user_id);
+		db_product = *(m_facade->GetRecommendation(user_id));
```
Comment (yaroslav-kernitskyi) [inline src/Server/Handlers/Product/GetRecommendationProductHandler.cpp]: Why not `.value()` ?

File: src/Server/Handlers/Product/GetRecommendationProductHandler.cpp [hunk 7]
```diff
@@ -27,22 +27,22 @@ Net::Response GetRecommendationProductHandler::AuthHandle(const Net::Request& re
...
 	}
 	catch (const db::NonexistentResource& ex) {
 		return FormErrorResponse(
 			NetError::Status::HTTP_NOT_FOUND,
 			"User with id = \"" + std::to_string(user_id) + "\" not found");
 	}
 
-	auto category = m_facade->GetProductCategoryById(db_product.category_id);
+	auto category = m_facade->GetProductCategoryById((*db_product).category_id);
```
Comment (yaroslav-kernitskyi) [inline src/Server/Handlers/Product/GetRecommendationProductHandler.cpp]: `(*object).field` can be replaced with `object->field`

File: src/Server/Handlers/Product/GetRecommendationProductHandler.cpp [hunk 8]
```diff
@@ -27,22 +27,22 @@ Net::Response GetRecommendationProductHandler::AuthHandle(const Net::Request& re
...
 	if (!category.has_value()) {
-			throw InternalError(std::string("No product category with id:") + std::to_string(db_product.category_id));
+			throw InternalError(std::string("No product category with id:") + std::to_string((*db_product).category_id));
```
Comment (yaroslav-kernitskyi) [inline src/Server/Handlers/Product/GetRecommendationProductHandler.cpp]: Same here

**Review by yaroslav-kernitskyi — APPROVED** [review]

**Review by VasylDobrovolskyi — APPROVED** [review]

**Pull request merged**